### DISCUSSION TITLE
Data fetcher not being invoked, query fails with non-null field is null

### DISCUSSION THREAD

**Discussion by shuston** [Netflix/dgs-framework#648] [Q&A] [answered]
Example reproducer for this is at: https://github.com/shuston/dgs-example

I have a schema with a simple interface, and a type implementing it and a query to return these things:
```
schema {
  query: Query
}

type ItemRange {
    name: String!
    first: Int!
    last: Int!
}

interface Group {
    label: String!
}

type DefinedGroup implements Group {
    label: String!
    items: [ItemRange!]!
}

type Query {
    groups: [Group!]!
}

```

My query handler returns an object without the fields of the DefinedGroup filled in:
```
@DgsComponent
public class GroupQuery {

	@DgsQuery
	public List<Group> groups() {
		DefinedGroup group = new DefinedGroup();
		group.setLabel("group-id");
		return Arrays.asList(group);
    }

}
```

And then I'm expecting a resolver for the 'items' field to be called:
```
@DgsComponent
public class DefinedGroupResolver {

	@Autowired
	private GroupService groupService;

	@DgsData(parentType = "DefinedGroup", field = "items")
	List<ItemRange> items(DgsDataFetchingEnvironment dfe) {
		DefinedGroup group = dfe.getSource();
		return groupService.findItemsForGroup(group.getLabel());
	}
}
```

However, this resolver for items is not called, and the query fails with:

```
com.netflix.graphql.dgs.exceptions.QueryException: The field at path '/groups[0]/items' was declared as a non null type, but the code involved in retrieving data has wrongly returned a null value.  The graphql specification requires that the parent field be set to null, or if that is non nullable that it bubble up null to its parent and so on. The non-nullable type is '[ItemRange!]' within parent type 'DefinedGroup'
	at com.example.GroupQueryTest.allGroups(GroupQueryTest.java:44)
```

Am I misunderstanding how this is supposed to work? Any suggestions for making it work?

Thank you,
-Steve

**Comment by paulbakker** [comment, marked as answer]
Hi @shuston, the reason is that your `items` datafetcher was missing the `public` keyword.
This works: 

```
@DgsData(parentType = "DefinedGroup", field = "items")
	public 	List<ItemRange> items(DgsDataFetchingEnvironment dfe) {
		DefinedGroup group = dfe.getSource();
		return groupService.findItemsForGroup(group.getLabel());
	}
```

I have to admit that it took me quite a bit of head-scratching to find this, it's really not obvious. I wasn't really aware the `public` is required, and nobody has complained so far. I'll create an issue to fix this.

**Reply by shuston** [reply]
`<forehead-slap>`
Thank you very, very much @paulbakker ... it works now.

**Comment by paulbakker** [comment]
And the issue as promised: https://github.com/Netflix/dgs-framework/issues/656
